### PR TITLE
filestate: Add and use re-usable worker pool

### PR DESCRIPTION
**Pull request by abhinav**
The filestate backend's Upgrade method currently
does some manual goroutine management
to ensure that it uses a fixed number of goroutines
as it attempts to upgrade all stacks in parallel.

This was fine while the upgrade step was just one phase:
Roughly:

    for _, stack := range stacks {
        go upgrade(stack)
    }
    // Using a pool instead of a new goroutine for each upgrade.

However, this will not suffice with upcoming changes to address #12600
because the upgrade process will now have multiple phases:
gather information, fill missing information with a prompt, upgrade.
The first and last phases of this are parallelizable,
but the second phase must run in a blocking manner
because it expects user input.

Adding that to the existing concurrency pattern in Upgrade
will lead to code that is very difficult to read.

This change introduces a simple shared worker pool abstraction.
It operates like a mix of `sync.WaitGroup` and [errgroup][1].
Specifically, it supports:

- multiple `Wait` and `Enqueue` phases (like WaitGroup)
- functions that return errors (like errgroup)

[1]: https://pkg.go.dev/golang.org/x/sync@v0.1.0/errgroup

This makes it very easy to adapt code that looks like the following:

    wg := &errgroup.Group{}
    for _, x := range xs {
        x := x
        wg.Go(func() error { return f(x) })
    }
    if err := wg.Wait(); err != nil {
        return err
    }

    wg = &errgroup.Group{}
    for _, y := range ys {
        y := y
        wg.Go(func() error { return f(y) })
    }
    if err := wg.Wait(); err != nil {
        return err
    }

Into code that looks like this:

    pool := newWorkerPool(..)
    defer pool.Close()

    for _, x := range xs {
        x := x
        pool.Enqueue(func() error { return f(x) })
    }
    if err := pool.Wait(); err != nil {
        return err
    }

    for _, y := range ys {
        y := y
        pool.Enqueue(func() error { return f(y) })
    }
    if err := pool.Wait(); err != nil {
        return err
    }

The workerPool-based version looks similar,
but it'll spawn a fixed number of goroutines once at the start
and re-use them for all tasks.

In service to #12600

### CODE REVIEW

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13076** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13076" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/13076?utm_source=stack-comment).

File: pkg/backend/filestate/backend.go [hunk 1]
```diff
@@ -377,53 +376,27 @@ func (b *localBackend) Upgrade(ctx context.Context) error {
...
+				upgraded.Add(1)
 			}
-		}()
+			return nil
+		})
 	}
-	wg.Wait()
+
+	if err := pool.Wait(); err != nil {
+		// Unreachable: All errors are logged.
```
Comment (Frassle) [inline pkg/backend/filestate/backend.go:396]: Should we `contract.Failf` here?

**Review by Frassle — APPROVED** [review]

**Comment by abhinav** [comment]
Failures are known failing codegen tests. Landing.

**Comment by abhinav** [comment]
bors merge

**Pull request merged**